br/pkg/lightning/importer/dup_detect.go

```diff
@@ -0,0 +1,387 @@
+	if err := parser.ReadRow(); err != nil {
+		if errors.Cause(err) == io.EOF {
+			return nil
+		}
+		return err
+	}
+	columnNames := parser.Columns()
+
+	cfg := d.rc.cfg
+	tblInfo := d.tr.tableInfo.Core
+
+	// 1. Initialize the column permutation.
+	igCols, err := d.rc.cfg.Mydumper.IgnoreColumns.GetIgnoreColumns(d.tr.dbInfo.Name, tblInfo.Name.O, cfg.Mydumper.CaseSensitive)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	ignoreColsMap := igCols.ColumnsMap()
+	colPerm, err := createColumnPermutation(columnNames, ignoreColsMap, tblInfo, d.logger)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// 2. Initialize the extend column values.
+	var extendVals []types.Datum
+	if len(chunk.FileMeta.ExtendData.Columns) > 0 {
+		_, extendVals = filterColumns(columnNames, chunk.FileMeta.ExtendData, ignoreColsMap, tblInfo)
+	}
+	lastRow := parser.LastRow()
+	lastRowLen := len(lastRow.Row)
+	extendColsMap := make(map[string]int)
+	for i, c := range chunk.FileMeta.ExtendData.Columns {
+		extendColsMap[c] = lastRowLen + i
+	}
+	for i, col := range tblInfo.Columns {
+		if p, ok := extendColsMap[col.Name.O]; ok {
+			colPerm[i] = p
+		}
+	}
+
+	// 3. Simplify table structure and create kv encoder.
+	encTable, colPerm, err := d.simplifyTable(colPerm)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	kvEncoder, err := d.rc.encBuilder.NewEncoder(ctx, &encode.EncodingConfig{
+		SessionOptions: encode.SessionOptions{
+			SQLMode:   d.rc.cfg.TiDB.SQLMode,
+			Timestamp: chunk.Timestamp,
+			SysVars:   d.rc.sysVars,
+			// use chunk.PrevRowIDMax as the auto random seed, so it can stay the same value after recover from checkpoint.
+			AutoRandomSeed: chunk.Chunk.PrevRowIDMax,
+		},
+		Path:   chunk.Key.Path,
+		Table:  encTable,
+		Logger: d.logger,
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	dataKVs := d.rc.encBuilder.MakeEmptyRows()
+	indexKVs := d.rc.encBuilder.MakeEmptyRows()
+	c := keyspace.CodecV1
+	if d.tr.kvStore != nil {
+		c = d.tr.kvStore.GetCodec()
+	}
+	dataChecksum := verify.NewKVChecksumWithKeyspace(c)
+	indexChecksum := verify.NewKVChecksumWithKeyspace(c)
+
+	for {
+		lastRow := parser.LastRow()
+		lastRow.Row = append(lastRow.Row, extendVals...)
+
+		kvs, err := kvEncoder.Encode(lastRow.Row, lastRow.RowID, colPerm, offset)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		kvs.ClassifyAndAppend(&dataKVs, dataChecksum, &indexKVs, indexChecksum)
+
+		for _, kvPair := range kv.Rows2KvPairs(dataKVs) {
+			if err := adder.Add(kvPair.Key, kvPair.RowID); err != nil {
+				return err
+			}
+		}
+		for _, kvPair := range kv.Rows2KvPairs(indexKVs) {
+			if err := adder.Add(kvPair.Key, kvPair.RowID); err != nil {
+				return err
+			}
+		}
+		dataKVs.Clear()
+		indexKVs.Clear()
+
+		offset, _ = parser.Pos()
+		if err := parser.ReadRow(); err != nil {
+			if errors.Cause(err) == io.EOF {
+				return nil
+			}
+			return err
+		}
+	}
+}
+
+func (d *dupDetector) simplifyTable(colPerm []int) (_ table.Table, newColPerm []int, _ error) {
```

Comment:
this function is not trivial, can you add a unit test for it?

Reply:
Addressed in https://github.com/pingcap/tidb/pull/42647/commits/27d4c248b060cc7cb4a1c27a876ce34c4c232aa3.